### PR TITLE
Feature/newapisystem

### LINKED ISSUE #1333: [Feature] Manage / Create API key + Manage IP access API
### In a few words, please describe the issue that you're experiencing:
Add options to manage API keys (Add / Revoke)
Manage IP access to API

+ Votes for adding certain access rights
Example:
1. Manage 1 account (Example user1)
2. Manage users account (Example user1, user2, ect...)
3. Manage Admin Settings 

Allow API access only to certain IP's

### PR DESCRIPTION
**Pull request by jhmaverick**
New API system with permission control and use with non-admin users.

Differences in handling keys:

* Each key is composed of an identification hash and a secret hash;
* Keys are linked with APIs and each API defines the scripts that can be used by it;
* To avoid conflicts with the keys of the old version, the new version is saved in "data/access-keys" and they have different management commands.
* API options are under "data/api".
* Non-admin users can use the APIs with limitations:
    * Scripts that have the "user" argument can only be used with the key's user;
    * Only APIs with ROLE "user" can be selected when creating a key;
    * Suspended users cannot consume from the API;
    * Key usage logs are also saved for the user;
    * The use of APIs by users can be disabled in the settings.
* Management of keys through the Web interface;
* Separate options to enable API versions in settings (It is possible to enable only for admin or for all users).

What has changed in the way of consuming the APIs:

* To make the connection it is necessary to replace the index "hash" by the indexes "access_key" and "secret_key" or send both in the index "hash" separated by 2 dots. Ex: "access_key:secret_key"(The system will define which API should be used by the 2 dots);
* A header called "Hestia-Exit-Code" with the code returned by the command is launched on requests;
* API will return a different http_code when an error occurs and for successful commands with empty output return "204";
* The "application/json" header is returned when the command returns a JSON.

### CODE REVIEW

File: bin/v-list-apis [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+# shellcheck source=/usr/local/hestia/func/main.sh
+source $HESTIA/func/main.sh
+# load config file
+source_conf "$HESTIA/conf/hestia.conf"
+
+# JSON list function
+json_list() {
+    echo '{'
+    local quote COMMANDS_ARR
+    for api in $(ls $HESTIA/data/api/); do
```
Comment (jaapmarcus) [inline bin/v-list-apis]: Error in Shell check 
8221	               ^---------------------^ SC2045 (error): Iterating over ls output is fragile. Use globs.

File: bin/v-list-apis [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    done
+    echo -e '\n}'
+}
+
+# SHELL list function
+shell_list() {
+    list="API\tROLE\tCOMMANDS\n"
+    list+="---\t----\t--------\n"
+
+    for api in $(ls $HESTIA/data/api/); do
```
Comment (jaapmarcus) [inline bin/v-list-apis]: Same here

File: bin/v-list-access-keys [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+source $HESTIA/func/main.sh
+# load config file
+source_conf "$HESTIA/conf/hestia.conf"
+
+# JSON list function
+json_list() {
+    echo -n '{'
+    local quote=''
+    local PERMISSIONS_ARR ACCESS_KEY_ID
+    for key_file in $(ls $HESTIA/data/access-keys/); do
```
Comment (jaapmarcus) [inline bin/v-list-access-keys]: Same here

File: bin/v-list-access-keys [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    [[ -n "$quote" ]] && echo
+    echo -e '}'
+}
+
+# SHELL list function
+shell_list() {
+    list="ID\tSECRET\tUSER\tPERMISSIONS\tCOMMENT\tTIME\tDATE\n"
+    list+="--\t------\t----\t-----------\t-------\t----\t----\n"
+
+    for key_file in $(ls $HESTIA/data/access-keys/); do
```
Comment (jaapmarcus) [inline bin/v-list-access-keys]: Same here

**Review by jaapmarcus — COMMENTED** [review]
See comment due to Shellcheck

Need to check PHP8.1 doesn't produce errors

**Review by ScIT-Raphael — APPROVED** [review]

File: install/hst-install-debian.sh [hunk 5]
```diff
@@ -2039,6 +2039,7 @@ write_config_value "POLICY_USER_VIEW_SUSPENDED" "no"
 write_config_value "POLICY_USER_VIEW_LOGS" "yes"
 write_config_value "POLICY_USER_EDIT_WEB_TEMPLATES" "true"
 write_config_value "POLICY_USER_EDIT_DNS_TEMPLATES" "yes"
+write_config_value "API_V2" "0"
```
Comment (jaapmarcus) [inline install/hst-install-debian.sh]: On Initial install it doesn't create the /usr/local/hestia/data/api key folder + copies the preset(s) to the folder

File: install/hst-install-ubuntu.sh [hunk 6]
```diff
@@ -2065,6 +2065,7 @@ write_config_value "POLICY_USER_VIEW_SUSPENDED" "no"
 write_config_value "POLICY_USER_VIEW_LOGS" "yes"
 write_config_value "POLICY_USER_EDIT_WEB_TEMPLATES" "true"
 write_config_value "POLICY_USER_EDIT_DNS_TEMPLATES" "yes"
+write_config_value "API_V2" "0"
```
Comment (jaapmarcus) [inline install/hst-install-ubuntu.sh]: Also applies for here

File: web/edit/server/index.php [hunk 7]
```diff
@@ -1045,6 +1045,19 @@
         }
     }
 
+    // Change API_V2
```
Comment (jaapmarcus) [inline web/edit/server/index.php]: Current javascript hides the ip box when disabling legacy api. This should not happen.

File: web/templates/pages/edit_server.html [hunk 8]
```diff
@@ -969,6 +969,21 @@
...
+                                                            <option value='0' <?php if(empty($_SESSION['API_V2']) || $_SESSION['API_V2'] == '0') echo 'selected' ?>><?=_('Disabled'); ?></option>
+                                                            <option value='1' <?php if($_SESSION['API_V2'] == '1') echo 'selected' ?>><?=_('Enabled for admin'); ?></option>
+                                                            <option value='2' <?php if($_SESSION['API_V2'] == '2') echo 'selected' ?>><?=_('Enabled for all users'); ?></option>
+                                                        </select>
+                                                        <br><br>
+                                                    </td>
+                                                </tr>
 												<tr>
 													<td>
 														<table id="security_ip" style="<?php if ($_SESSION['API'] == "no") echo 'display:none;';?>">
```
Comment (jaapmarcus) [inline web/templates/pages/edit_server.html]: IP box should not get hidden after Legacy is disabled

File: web/templates/pages/add_access_key.html [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                        } ?>
+
+                        <tr>
+                            <td class="vst-text step-top">
+                                <?= _('Comment'); ?> <span class="optional">(<?= _('optional'); ?>)</span>
+                            </td>
+                        </tr>
+                        <tr>
+                            <td>
+                                <input type="text" size="20" class="vst-input" name="v_comment" maxlength="255"
```
Comment (jaapmarcus) [inline web/templates/pages/add_access_key.html]: 2022/04/18 19:08:03 [error] 1012285#0: *30 FastCGI sent in stderr: "PHP message: PHP Warning:  Undefined variable $v_comment in /usr/local/hestia/web/templates/pages/add_access_key.html on line 72PHP message: PHP Deprecated:  trim(): Passing null to parameter #1 ($string) of type string is deprecated in /usr/local/hestia/web/templates/pages/add_access_key.html on line 72" while reading response header from upstream,

File: web/templates/pages/list_access_key.html [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+                        <?php } ?>
+
+                        <?php
+                        if (!empty($_SESSION['ok_msg'])) {
+                            ?>
+                            <?php if (!empty($key_data['ACCESS_KEY_ID']) && !empty($key_data['SECRET_ACCESS_KEY'])) { ?>
+                                <tr>
+                                    <td class="vst-text input-label">
+                                        <?= _('Secret Key'); ?><br>
+                                        <span style="font-size: 10pt; color:#ffd500;"><?= _('WARNING_TO_SAVE_ACCESS_KEY'); ?></span>
```
Comment (jaapmarcus) [inline web/templates/pages/list_access_key.html]: Screenshot 2022-04-18 at 19 10 08

With the dark theme it is hard to read we should probally change the color slightly

**Review by jaapmarcus — COMMENTED** [review]
It is missing an upgrade "path" from non existing setup to existing in 1.6.0.sh

File: install/hst-install-debian.sh (hunk 5 above)
Comment (jaapmarcus) [reply install/hst-install-debian.sh]: We should probally enable it for admin user only on default if api is enabled

File: install/hst-install-debian.sh [hunk 11]
```diff
@@ -1869,7 +1872,8 @@ $HESTIA/bin/v-add-sys-phpmailer quiet
 #----------------------------------------------------------#
 
 if [ "$api" = "yes" ]; then
-    write_config_value "API" "yes"
+    write_config_value "API" "no"
```
Comment (jaapmarcus) [inline install/hst-install-debian.sh]: Untill we have fully finished the transition transition we should keep it enabled. for now

File: install/hst-install-ubuntu.sh [hunk 12]
```diff
@@ -1915,7 +1918,8 @@ fi
 #----------------------------------------------------------#
 
 if [ "$api" = "yes" ]; then
-    write_config_value "API" "yes"
+    write_config_value "API" "no"
```
Comment (jaapmarcus) [inline install/hst-install-ubuntu.sh]: Untill we have fully finished the transition transition we should keep it enabled. for now

**Review by jaapmarcus — CHANGES_REQUESTED** [review]

**Comment by jaapmarcus** [comment]
Disabling legacy api "removed" api completely causing to break  the "new" api. 

Changes pushed solved the issue

**Comment by jhmaverick** [comment]
Yesterday I didn't have time to check what needed to be changed in the edit/server index. But I think now everything is finished.

**Review by jaapmarcus — APPROVED** [review]

**Comment by vikasprogrammer** [comment]
Do we have a documentation entry for this?

**Comment by jaapmarcus** [comment]
https://github.com/hestiacp/hestiacp-docs/commit/b0e8a27f672e6af57197fb8e12369535690beb70

I push some changes to GitHub but for unknown reasons they are not merged yet on the live site

**Comment by jaapmarcus** [comment]
https://docs.hestiacp.com/admin_docs/api.html#api-key-vs-password-vs-access-key

**Pull request merged**